Sinatra practiceの提出物です

rira100000000 opened the pull request:
[Sinatra を使ってWebアプリケーションの基本を理解する \| FBC](https://bootcamp.fjord.jp/practices/157)の提出物です。

rira100000000 committed:
主な機能の実装

新規登録機能を実装しました
一覧表示機能を実装しました
表示機能を実装しました
編集機能を実装しました
削除機能を実装しました

rira100000000 committed:
削除のバグを修正

削除時に1つ前のインデックスのファイルを削除するバグを修正

rira100000000 committed:
rubocop対応

rubocopの指摘部分の修正をしました

rira100000000 committed:
erb_lintの対応

erb_lintの指摘箇所を修正しました

rira100000000 committed:
URIの修正

memos/edit/1という形からmemos/1/editの形に変更しました

rira100000000 committed:
XSS攻撃対応

フォームをXSS攻撃に対応させました

rira100000000 committed:
layout.erbの追加

layout.erbを追加しました
viewsの重複する処理を削除しました

rira100000000 committed:
インデントの修正

rira100000000 committed:
CSS対応

memo.cssを追加しました
viewsの各ファイルをCSSに対応するよう修正しました

rira100000000 committed:
README.mdの更新

rira100000000 committed:
データファイルに拡張子を追加

データファイルのファイル名を"1"から"1.txt"の形に変更しました

rira100000000 committed:
不要なファイルの削除

rira100000000 committed:
トップページの設定

http://localhost:4567にアクセスした時、http://localhost:4567/memosにリダイレクトするよう修正しました
READMEにhttp://localhost:4567にアクセスするよう記載しました

rira100000000 committed:
rubocop対応

cafedomancer commented on README.md:
```diff
@@ -0,0 +1,33 @@
+# Note
+作成したメモのデータはdataディレクトリ配下にIDを名前として保存されます。<br>
+作成したメモはrootディレクトリのfile_infos.txtで管理されます。
+‘bundle exec ruby app.r`bundle exec ruby app.rb`実行後、ブラウザから`http://localhost:4567`にアクセスしてください。
```
文章がなにかおかしくなっていそうです。

rira100000000 replied:
修正します。

rira100000000 committed:
dataディレクトリの生成処理を追加

dataディレクトリがない場合、自動で作成するように修正しました

rira100000000 committed:
データの保存ファイルをmemos.csvに変更

メモ一覧表示時にfile_infos.txtを使用しないように変更しました
メモ新規作成時にfile_infos.txtを使用しないように変更しました
メモ表示時にfile_infos.txtを使用しないように変更しました

rira100000000 committed:
edit,update処理でfile_infos.txtをしないように変更

fetch_title_and_contentの名前をfetch_memoに修正しました
CSVに対してforeachするとき、取り出した要素名をmemoに統一しました
memos.csvにヘッダー情報を追加しました
csvファイルに書き込む際に"で囲むよう修正しました

rira100000000 committed:
delete処理の修正

delete処理でfile_info.txtを使用しないように修正しました
delete処理をフラグ管理から実際にデータを削除するよう修正しました

rira100000000 committed:
不要なメソッドの削除

使わなくなったメソッドreplace_lineを削除しました

rira100000000 committed:
リファクタリング

fetch_memoメソッドでidをparamsから取得するようにしました
:strict_pathsを設定し/の有無でルーティングが変わわないように修正しました

rira100000000 committed:
max_id.txtの作成処理追加

max_id.txtがない場合、作成するように修正しました

rira100000000 committed:
rubocop対応

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+require 'sinatra/reloader'
+require 'csv'
+
+set :strict_paths, false
```
👍

rira100000000 replied:
😄

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+  if File.exist?("#{data_dir}/memos.csv")
+    @memos = []
+    CSV.foreach("#{data_dir}/memos.csv", headers: true) do |memo|
+      @memos << [memo[0], memo[1]]
```
`CSV.foreach` に `headers: true` を指定して `CSV::Row` オブジェクトが生成されるようにしているのに、ここでインデックスアクセスしてしまうとヘッダーを設定している意味がなくなってしまう気がしました。

rira100000000 replied:
memo[0]→memo['id']のようにアクセスするよう修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+  data_dir = "#{Dir.pwd}/data"
+  Dir.mkdir(data_dir) unless Dir.exist?(data_dir)
+  prepare_max_id_file(data_dir)
+  if File.exist?("#{data_dir}/memos.csv")
```
`"#{data_dir}/memos.csv"` という文字列があちこちで重複しています。ファイル名を変更するときにコードを書き換えてまわらないといけなくて大変なので、どこか一箇所で定義するようにしたほうが良さそうです。

rira100000000 replied:
パスの指定をbeforeメソッド内で行うよう修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+      @memos << [memo[0], memo[1]]
+    end
+  else
+    file = File.new("#{data_dir}/memos.csv", 'w')
+    file.puts('"id","title","content"')
+    file.close
```
`close` の実行し忘れを防ぐために、`File.open` とブロックを組み合わせ処理するようにしたほうが良さそうです。また、`data` ディレクトリに関するコメントと同様に、この処理も `/memos` に最初にアクセスされることが暗黙の前提条件となっていそうです。`data` ディレクトリの作成と合わせて、実装を見直してみてください。また、これは CSV ファイルの処理なので `csv` ライブラリを使って行ったほうがいいかなと思いました。

rira100000000 replied:
ファイルの存在確認は`prepare_data_file`メソッドに切り出して`before`内で行うよう修正します。
CSVライブラリとブロックを用いた処理に修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+  erb :new
+end
+
+post '/memos/create' do
```
RESTful な URL としては `POST /memos` でメモが作成されることを表現できているので、`/create` はいらないかなーと思います。`GET /`, `GET /memos`, `GET /memos/new` を除く、他のルーティングについても同様です。

rira100000000 replied:
URLの`/create`、`/show`、`/update`、`/delete`部分を削除します。
`get '/memos/:id/edit'`については必要に感じたので残しますが、問題があればコメントお願いします。

cafedomancer replied:
すみません、edit は必要でした 🙇

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+require 'csv'
+
+set :strict_paths, false
+enable :method_override
```
この設定は明示的に行わないと有効にならないでしょうか？ Sinatra のドキュメントを確認してみてください。

rira100000000 replied:
[Sinatra: Configuring Settings](https://sinatrarb.com/configuration.html)
上記で`method_override`の項を確認したのですが、デフォルト値についての言及はありませんでした。
`settings.method_override`にてデフォルト値を調べたところ`true`だったため、`enable :method_override`は削除します。

cafedomancer replied:
Sinatra の README に記述があるはずなので確認してみてください。https://sinatrarb.com/intro.html

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+  index = table.each_with_index do |row, i|
+    break i if row[:id] == id
+  end
+  table[index] = [id.to_s, title.to_s, content.to_s]
```
質問: `title` と `content` を `to_s` している理由を教えてください。

rira100000000 replied:
titleやcontentに数値が入った場合を考慮した記憶がありますが、特に必要がなかったので修正します。

cafedomancer commented on views/layout.erb:
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html lang='ja'>
```
属性値の指定にシングルクォーテーションが使われていたりダブルクォーテーションが使われたりしています。どちらかに統一しましょう。ダブルクォーテーションが一般的かなと思います。

rira100000000 replied:
ダブルクォーテーションで囲むよう修正します。

cafedomancer commented on views/layout.erb:
```diff
@@ -0,0 +1,14 @@
+</head>
+<body>
+<header>rira's memo app</header>
+  <div class=container>
```
属性値はダブルクォーテーションで囲みましょう。

rira100000000 replied:
ダブルクォーテーションで囲むよう修正します。

cafedomancer commented on views/layout.erb:
```diff
@@ -0,0 +1,14 @@
+  <title><%= @page_title %></title>
+</head>
+<body>
+<header>rira's memo app</header>
```
この行のインデントが整っていません。

rira100000000 replied:
インデントの修正をします。

cafedomancer commented on views/memos.erb:
```diff
@@ -0,0 +1,9 @@
+<div class=mid_box>
+  <a href='http://localhost:4567/memos/new' class=button>新規作成</a><br>
```
プロトコルやホストやポートが変わると、この href 属性の値ではリンクが正しく動かなくなってしまうんじゃないでしょうか。

rira100000000 replied:
URLの指定時は"/memos/new"の形で指定するよう修正します。

cafedomancer commented on views/memos.erb:
```diff
@@ -0,0 +1,9 @@
+  <% unless @memos.nil? %>
+    <% @memos.each do |memo| %>
+      <% id, title = *memo %>
+      <a href="<%= "http://localhost:4567/memos/#{id}/show" %>"> <%= "●" + title %> </a><br>
```
`"●"` 部分を `<%= ... %>` の中に含める必要はない気がします。

cafedomancer commented on views/memos.erb:
```diff
@@ -0,0 +1,9 @@
+  <a href='http://localhost:4567/memos/new' class=button>新規作成</a><br>
+  <% unless @memos.nil? %>
+    <% @memos.each do |memo| %>
+      <% id, title = *memo %>
```
多重代入して説明的な変数を使うよりも、そもそものデータ構造を見直したほうが良さそうです。名前を持たせたいのであれば、配列よりもハッシュやオブジェクトなどのデータ構造のほうが適しているんじゃないでしょうか。

rira100000000 replied:
`memo`をハッシュに修正します。

「ハッシュを使うことで、（多重代入部分の）コードが1行減り、シンプルになるから」というのが私の理解なのですが、そもそも多重代入は積極的に使わない方がいいのでしょうか？

cafedomancer replied:
コードの行数を減らしたり、多重代入を避けたりしたいわけではないです。そもそも配列は、同じ種類のデータをまとめて扱ったり、順番に意味のあるデータを扱ったりするためのデータ構造です。今回の `memo` の `id` と `title` はそうではないので、配列を使うことが適していません。そのことが多重代入や、別のところだとインデックスアクセスの形で表れていたので、別のデータ構造のほうが適しているのでは、と判断しました。

rira100000000 replied:
そういうことだったのですね。
やっと理解できました。ありがとうございます。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+
+get '/memos/:id/show' do
+  fetched_memo = fetch_memo(params[:id])
+  @id = fetched_memo['id']
+  @title = fetched_memo['title']
+  @content = fetched_memo['content']
```
これらのデータをばらして ERB ファイルで使うよりも、メモというまとまりでそのまま使うといいんじゃないでしょうか。例えばですが、メモに id, title, content 以外の項目が増えたときには、その項目についてもここでインスタンス変数を追加で定義しなければならないのが理由です。

rira100000000 replied:
データはばらさず、`@memo['id']`のようにそのまま利用するよう修正します。

rira100000000 committed:
.gitignoreの修正

使われなくなったfile_infos.txtが含まれていたため削除しました

rira100000000 committed:
.rubocop.ymlを作成

rira100000000 committed:
READMEで重複する内容を削除

READMEのインストールの項を削除しました

rira100000000 committed:
CSSファイルのスタイルを統一

CSSファイルのスタイルが統一されていなかったため、修正しました

rira100000000 commented on README.md:
````diff
@@ -0,0 +1,32 @@
+
+clone後にbundle installを使って必要なgemをインストールできます。
+
+```bash
+bundle install
+```
````
インストールと使い方で内容が重複していたので、インストールの項を削除します。

rira100000000 commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+data/
+file_infos.txt
```
`file_infos.txt`は不要になったため削除します。

rira100000000 commented on public/memo.css:
```diff
@@ -0,0 +1,79 @@
+}
+
+.button{
+  display: inline;
+  margin-right: 20px;
+  margin-top: 20px;
+  height: 28px;
+  width: 75px;
+  border: 1px solid black;
+  border-radius: 15%;
+  background-color:  #a07ca3;
+  color: #bed92c;
+  text-decoration: none;
+  font-size: large;
+  text-align: center;
+  cursor: pointer;
+}
+
+#title{
+  width: 80%;
+  font-size:  1.4em;
+}
+
+#content{
+  width: 80%;
+  height: 300px;
+  text-align: left;
+  display: table-cell;
+  vertical-align: text-top;
+  font-size:  1.4em;
+}
+
+.title_area{
+  width: 100%;
+}
+.display_new_line {
+  white-space: pre-wrap;
+  word-wrap: break-word;
+  width: 100%;
+}
```
スタイルの修正をします。

rira100000000 committed:
誤記部分を修正

rira100000000 commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+end
+
+get '/' do
+  redirect redirect '/memos'
```
誤記でしたので削除します。

rira100000000 committed:
不要な処理の削除

デフォルトで有効になっていたため、enable :method_overrideを削除しました
必要のないto_s処理を削除しました

rira100000000 committed:
erbファイルの体裁修正

インデントを整えました
属性値はダブルクォーテーションで囲むよう統一しました

rira100000000 committed:
<%= ... %>にただの文字列を含めないように修正しました

rira100000000 commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+  data_dir = "#{Dir.pwd}/data"
+  title = protect_xss(params[:title])
+  content = protect_xss(params[:content])
+  id = protect_xss(params[:id]).to_i
```
hiddenなinputに対してもXSS攻撃ができると思っていたのですが、実際には`accesskey`属性が付与されていなければ攻撃できないということでした。
[hiddenなinput要素のXSSでJavaScript実行 \| 徳丸浩の日記](https://blog.tokumaru.org/2016/04/hiddeninputxssjavascript.html)

`params[:id]`へのXSS対策は行わないよう修正します。

rira100000000 committed:
不要なXSS対策を削除しました

rira100000000 committed:
beforeの追加

パスの指定をbefore内で行うように修正しました
必要なファイル、ディレクトリの生成処理をbefore内で行うように修正しました

rira100000000 commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+get '/memos' do
+  @page_title = 'メモ一覧'
+  data_dir = "#{Dir.pwd}/data"
+  Dir.mkdir(data_dir) unless Dir.exist?(data_dir)
```
どのURLにアクセスしても動作するように、`before`メソッド内で行うように修正します。

rira100000000 commented on app.rb:
```diff
@@ -0,0 +1,126 @@
+
+get '/memos' do
+  @page_title = 'メモ一覧'
+  data_dir = "#{Dir.pwd}/data"
```
パスの指定は`before`メソッド内で行うよう修正します。
パス結合では`File.join`を使うよう修正します。

rira100000000 committed:
prepare_data_fileメソッドの追加

prepare_max_id_fileに合わせて、データ保存用CSVの存在確認、作成はprepare_data_fileで行うように修正しました
csvファイルを開くときはcsvライブラリを利用するよう修正しました。

rira100000000 committed:
prepare_max_id_fileの修正

変数名を間違えていたので修正しました
File.openでブロックを使用するよう修正しました

rira100000000 committed:
CSV::Rowオブジェクトにアクセス方法の変更

CSV::Rowオブジェクトにアクセスする時はヘッダーを用いてアクセスするよう修正しました

rira100000000 committed:
URLの修正

URLの/create、/show、/update、/delete部分を削除しました

rira100000000 committed:
データ構造の変更

memoをハッシュに変更しました
memosが空か識別する際の処理をnil?からempty?に修正しました

rira100000000 committed:
fetch_memoした結果を分解しないよう修正

fetch_memoした結果を@memoに格納して、そのままviewに渡すように修正しました

rira100000000 committed:
Rubocop対応

rira100000000 committed:
メモ表示時にタイトルと本文の間に罫線を追加しました

rira100000000 committed:
ファイルのopen時はブロックを使うよう修正

File.openとCSV.open時にブロックを使うよう修正しました

rira100000000 committed:
空行の追加

コードの意味的なまとまりをわかりやすくするために空行を追加しました

cafedomancer commented on README.md:
````diff
@@ -0,0 +1,24 @@
+# sinatra_practice
+[Sinatra を使ってWebアプリケーションの基本を理解する \| FBC](https://bootcamp.fjord.jp/practices/157)の提出課題です。<br>
+sinatraを使って簡単なメモアプリを作成しました。
+
+# 必要なgem
+  
+* gem 'sinatra'
+* gem 'sinatra-contrib'
+* gem 'webrick'
+
+# 使い方
+
+```bash
+git clone -b sinatra_practice https://github.com/rira100000000/sinatra_practice.git
+cd <cloneしたディレクトリ>
+bundle install
+bundle exec ruby app.rb
+```
+ 
+# Note
+作成したメモのデータはdataディレクトリ配下にmemos.csvとして保存されます。<br>
+`bundle exec ruby app.rb`実行後、ブラウザから`http://localhost:4567`にアクセスしてください。
+# 作成者
+rira100000000
````
ちょっと細かいですが、`#` から始まる見出しの行の前後に空行があったりなかったりするので、何らかの形で統一すると良いと思います。

rira100000000 replied:
修正、統一します。

cafedomancer commented on public/memo.css:
```diff
@@ -0,0 +1,70 @@
+}
+.menu {
+  display: flex;
+}
+.button {
+  display: inline;
+  margin-right: 20px;
+  margin-top: 20px;
+  height: 28px;
+  width: 75px;
+  border: 1px solid black;
+  border-radius: 15%;
+  background-color: #a07ca3;
+  color: #bed92c;
+  text-decoration: none;
+  font-size: large;
+  text-align: center;
+  cursor: pointer;
+}
+#title {
+  width: 80%;
+  font-size: 1.4em;
+}
+#content {
+  width: 80%;
+  height: 300px;
+  text-align: left;
+  display: table-cell;
+  vertical-align: text-top;
+  font-size: 1.4em;
+}
+.title_area {
+  width: 100%;
+  border-bottom: 1px solid #79854c;
+}
+.display_new_line {
+  white-space: pre-wrap;
+  word-wrap: break-word;
+  width: 100%;
+}
```
見づらくなってしまうので宣言ブロック同士の間は空行を入れたほうがよいです。

rira100000000 replied:
空行を追加します。

cafedomancer commented on public/memo.css:
```diff
@@ -0,0 +1,70 @@
+  text-decoration: none;
+  font-size: 1.3m;
+}
+h1 {
+  color: #a7a7a7;
+  margin-block-start: 0;
+  margin-block-end: 0;
+  font-size: 1.7em;
+}
```
h1 タグはどこでも使われていない気がしました。

rira100000000 replied:
修正するうちに不要になっていました。削除します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,118 @@
+set :strict_paths, false
+
+before do
+  data_dir = File.join(Dir.pwd, 'data')
+  @csv_path = File.join(data_dir, 'memos.csv')
+  @max_id_path = File.join(data_dir, 'max_id.txt')
```
これらの値は変わらない値なので定数にしておくのがよいかなーと思います。

rira100000000 replied:
それぞれ定数化します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,118 @@
+  prepare_data_file(@csv_path)
+end
+
+get '/memos' do
+  @page_title = 'メモ一覧'
+  @memos = []
+
+  CSV.foreach(@csv_path, headers: true) do |memo|
+    @memos << { id: memo['id'], title: memo['title'] }
+  end
+  erb :memos
+end
```
意味的に以下のようなまとまりにするのがよいと思います。コメントは補足のためのものなので残さなくてよいです。

```suggestion
get '/memos' do
  # メモの取得
  @memos = []
  CSV.foreach(@csv_path, headers: true) do |memo|
    @memos << { id: memo['id'], title: memo['title'] }
  end

  # ページの表示
  @page_title = 'メモ一覧'
  erb :memos
end
```

rira100000000 replied:
同様の箇所も含めて、そのように修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,118 @@
+  erb :new
+end
+
+post '/memos' do
+  title = protect_xss(params[:title])
+  content = protect_xss(params[:content])
+  id = File.read(@max_id_path).to_i + 1
+
+  CSV.open(@csv_path, 'a', quote_char: '"') { |csv| csv << [id, title, content] }
+  File.open(@max_id_path, 'w') { |file| file << id }
+  redirect "/memos/#{id}"
+end
```
引数の使われ方や意味のまとまり的に以下のようにするのがよいかなと思います。

```suggestion
post '/memos' do
  # メモの準備
  id = File.read(@max_id_path).to_i + 1
  title = protect_xss(params[:title])
  content = protect_xss(params[:content])

  # メモの保存
  CSV.open(@csv_path, 'a', quote_char: '"') { |csv| csv << [id, title, content] }
  File.open(@max_id_path, 'w') { |file| file << id }

  # 画面の遷移
  redirect "/memos/#{id}"
end
```

rira100000000 replied:
そのように修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,118 @@
+
+get '/memos/:id' do
+  @memo = fetch_memo(params[:id], @csv_path)
+  @page_title = @title
```
`@title` は宣言されていない気がします。

rira100000000 replied:
@memo['title']の誤りでした。修正します。

rira100000000 replied:
`@memo['title']`の誤りでした。修正します。

cafedomancer commented on views/edit.erb:
```diff
@@ -0,0 +1,9 @@
+  <label for="content">本文</label><br>
+  <textarea type="text" id="content" name="content"><%= @memo['content'] %></textarea><br>
+  <input type="submit" value="変更" class="button">
+  <input type="hidden" id="id" name="id" value="<%= @memo['id'] %>">
```
id は action 属性のパスに含まれているので、hidden フィールドで指定する必要はない気がしました。

rira100000000 replied:
不要でしたので、削除します。

rira100000000 committed:
READMEの体裁修正

見出し前後の空行の有無を統一しました。

rira100000000 committed:
CSSファイルの修正

宣言同士の間に空行を追加しました
h1タグは使用していないのでっ削除しました

rira100000000 committed:
app.rbの体裁修正

意味のまとまりがわかりやすくなるように空行を入れました
処理の順番をid,title,contentの順に統一しました

rira100000000 committed:
ページタイトルのバグ修正

メモ表示、メモ編集画面でページタイトルが取得できないバグを修正しました

rira100000000 committed:
不要なinputの削除

idはaction属性のパスに含まれているため、削除しました

rira100000000 committed:
ファイル名、パスの定数化

ファイル、ディレクトリ名とパスを定数化しました

rira100000000 committed:
ハッシュのキーをシンボルに統一しました

rira100000000 committed:
rubocop対応

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,129 @@
+get '/memos' do
+  @memos = []
+  CSV.open(CSV_PATH, headers: true) do |csv|
+    csv.header_convert(:symbol)
```
この行は `CSV.open` のオプションとして指定する方法があるので調べてみてください。`CSV.open` は `CSV.new` と同じオプションを受け取るようになっています。
https://docs.ruby-lang.org/ja/latest/method/CSV/s/new.html
https://docs.ruby-lang.org/ja/latest/method/CSV/s/open.html

rira100000000 replied:
header_convertersオプションがありましたので、そちらを使うよう修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,129 @@
+  content = protect_xss(params[:content])
+
+  table = CSV.table(CSV_PATH)
+  index = table.each_with_index do |row, i|
+    break i if row[:id] == id
+  end
```
もしかしたらここは `Enumerable#find_index` を使うとスッキリ書けるかもしれません。https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/find_index.html

rira100000000 replied:
find_indexで処理するよう修正します。

cafedomancer commented on app.rb:
```diff
@@ -0,0 +1,129 @@
+def fetch_memo(id)
+  CSV.open(CSV_PATH, headers: true) do |csv|
+    csv.header_convert(:symbol)
+    csv.each do |memo|
+      return memo if memo[:id] == id
+    end
```
ここは `Enumerable#find` が使えるかもしれません。https://docs.ruby-lang.org/ja/latest/class/Enumerable.html#I_DETECT

rira100000000 replied:
findを使用するよう修正します。

rira100000000 committed:
header_convertersを使用するよう修正しました

rira100000000 committed:
find_indexを使用するよう修正

rira100000000 committed:
Enumerable#findを使用するよう修正

cafedomancer reviewed (approved):
🌟
